unsafe_split, unsafe_split_with_sizes, unsafe_chunk operations

Baranowski opened the pull request:
Fixes #36403

Copy-paste of the issue description:

* Escape hatch: Introduce unsafe_* version of the three functions above that have the current behavior (outputs not tracked as views). The documentation will explain in detail why they are unsafe and when it is safe to use them. (basically, only the outputs OR the input can be modified inplace but not both. Otherwise, you will get wrong gradients).
* Deprecation: Use the CreationMeta on views to track views created by these three ops and throw warning when any of the views is modified inplace saying that this is deprecated and will raise an error soon. For users that really need to modify these views inplace, they should look at the doc of the unsafe_* version to make sure their usecase is valid:
  * If it is not, then pytorch is computing wrong gradients for their use case and they should not do inplace anymore.
  * If it is, then they can use the unsafe_* version to keep the current behavior.
* Removal: Use the CreationMeta on view to prevent any inplace on these views (like we do for all other views coming from multi-output Nodes). The users will still be able to use the unsafe_ versions if they really need to do this.


Note about BC-breaking:
- This PR changes the behavior of the regular function by making them return proper views now. This is a modification that the user will be able to see.
- We skip all the view logic for these views and so the code should behave the same as before (except the change in the `._is_view()` value).
- Even though the view logic is not performed, we do raise deprecation warnings for the cases where doing these ops would throw an error.

Baranowski committed:
Squashed

Baranowski commented:
Note: when using the safe (i.e., original) versions of those functions, modifying a resulting view inplace causes only a warning. But calling `backward()` still causes a `RuntimeError`:

```
In [1]: import torch

In [2]: a = torch.randn(3,3,requires_grad=True)

In [3]: s1,s2,s3 = a.chunk(3)

In [4]: s1.mul_(s2)
/home/wbaranowski/miniconda3/envs/pytorch-cuda-dev/bin/ipython:1: UserWarning: Output 0 of SplitBackward is a view and is being modified inplace. This view is the output of a function that returns multiple views. Inplace operators on such views are being deprecated and soon will not be allowed. Consider using `unsafe_` version of the function that produced this view. (Triggered internally at  ../torch/csrc/autograd/variable.cpp:469.)
  #!/home/wbaranowski/miniconda3/envs/pytorch-cuda-dev/bin/python
Out[4]: tensor([[-0.1466, -0.8976,  0.4805]], grad_fn=<AsStridedBackward>)

In [5]: s1.sum().backward()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-529ea6b4085e> in <module>
----> 1 s1.sum().backward()

~/git/Quansight/pytorch/torch/tensor.py in backward(self, gradient, retain_graph, create_graph)
    182                 products. Defaults to ``False``.
    183         """
--> 184         torch.autograd.backward(self, gradient, retain_graph, create_graph)
    185 
    186     def register_hook(self, hook):

~/git/Quansight/pytorch/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables)
    123     Variable._execution_engine.run_backward(
    124         tensors, grad_tensors, retain_graph, create_graph,
--> 125         allow_unreachable=True)  # allow_unreachable flag
    126 
    127 

RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [torch.FloatTensor [1, 3]], which is output 1 of SplitBackward, is at version 1; expected version 0 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True).
```

albanD commented:
| model | task | mean (before) | var (before) | mean (after) | var (after) | speedup |
| resnet18 | vjp | 5.245208853921213e-07 | 3.132704057100255e-13 | 3.099441414633475e-07 | 2.5326900407636466e-13 | 1.6923077910609536 |
| resnet18 | vhp | 5.245208853921213e-07 | 3.132704057100255e-13 | 3.814697322468419e-07 | 2.5516378289805464e-13 | 1.3750000093132255 |
| resnet18 | jvp | 2.8610230629055877e-07 | 2.2484729464596065e-13 | 1.9073486612342094e-07 | 1.6168795089702842e-13 | 1.5000000372529025 |
| fcn_resnet | vjp | 3.099441414633475e-07 | 5.059064061771479e-13 | 2.8610230629055877e-07 | 2.1221543131718507e-13 | 1.0833332505491078 |
| fcn_resnet | vhp | 3.814697322468419e-07 | 2.425319331218062e-13 | 3.814697322468419e-07 | 2.5516378289805464e-13 | 1.0 |
| fcn_resnet | jvp | 5.960464477539062e-07 | 4.831690603168681e-13 | 5.960464477539062e-07 | 5.336965678420791e-13 | 1.0 |
| ppl_simple_reg | vjp | 3.099441414633475e-07 | 2.5326900407636466e-13 | 6.437301749429025e-07 | 3.1642836137782404e-13 | 0.4814814553175775 |
| ppl_simple_reg | vhp | 2.1457671550706436e-07 | 2.0779427169822368e-13 | 5.722046125811175e-07 | 2.5516378289805464e-13 | 0.3749999751647323 |
| ppl_simple_reg | jvp | 0.0 | 0.0 | 5.722046125811175e-07 | 2.425319331218062e-13 | 0.0 |
| ppl_simple_reg | hvp | 5.722046125811175e-07 | 2.5516378289805464e-13 | 4.0531159584134e-07 | 2.7853275783897014e-13 | 1.411764722381908 |
| ppl_simple_reg | jacobian | 0.0 | 0.0 | 4.76837158203125e-07 | 2.6526927898208597e-13 | 0.0 |
| ppl_simple_reg | hessian | 2.8610230629055877e-07 | 2.2484729464596065e-13 | 4.76837158203125e-07 | 2.6526927898208597e-13 | 0.6000000238418579 |
| ppl_robust_reg | vjp | 3.099441414633475e-07 | 2.5326900407636466e-13 | 3.099441414633475e-07 | 2.5326900407636466e-13 | 1.0 |
| ppl_robust_reg | vhp | 5.006789933759137e-07 | 2.8358549232845864e-13 | 1.9073486612342094e-07 | 1.6168795089702842e-13 | 2.6249998416751645 |
| ppl_robust_reg | jvp | 3.099441414633475e-07 | 2.5326900407636466e-13 | 2.8610230629055877e-07 | 2.2484729464596065e-13 | 1.0833332505491078 |
| ppl_robust_reg | hvp | 3.814697322468419e-07 | 2.425319331218062e-13 | 3.5762786865234375e-07 | 2.17899767782255e-13 | 1.0666666825612385 |
| ppl_robust_reg | jacobian | 3.099441414633475e-07 | 2.5326900407636466e-13 | 4.2915343101412873e-07 | 3.132704057100255e-13 | 0.7222222148636239 |
| ppl_robust_reg | hessian | 1.9073486612342094e-07 | 1.6168795089702842e-13 | 2.8610230629055877e-07 | 2.1221543131718507e-13 | 0.6666666501098215 |
| wav2letter | vjp | 5.245208853921213e-07 | 3.132704057100255e-13 | 4.76837158203125e-07 | 4.673792277677669e-13 | 1.100000023841858 |
| wav2letter | vhp | 2.8610230629055877e-07 | 2.1221543131718507e-13 | 3.337860050578456e-07 | 2.9305941354196285e-13 | 0.8571429057997109 |
| wav2letter | jvp | 4.0531159584134e-07 | 2.7853275783897014e-13 | 6.19888282926695e-07 | 2.9305941354196285e-13 | 0.6538461961689801 |
| deepspeech | vjp | 3.814697322468419e-07 | 2.425319331218062e-13 | 3.337860050578456e-07 | 2.9305941354196285e-13 | 1.142857179349783 |
| transfo | vjp | 3.099441414633475e-07 | 2.5326900407636466e-13 | 3.099441414633475e-07 | 2.5326900407636466e-13 | 1.0 |
| transfo | vhp | 3.099441414633475e-07 | 2.5326900407636466e-13 | 3.814697322468419e-07 | 2.425319331218062e-13 | 0.8124999580904848 |
| transfo | jvp | 4.0531159584134e-07 | 2.7853275783897014e-13 | 4.76837158203125e-07 | 4.168517473476102e-13 | 0.8500000238418579 |

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -341,7 +341,6 @@ const std::shared_ptr<torch::autograd::Node>& VariableHooks::grad_fn(const Tenso
     if (diff_view_meta->attr_version != current_version) {
       // This is an indirect rebase_history due to another view or the base being modified inplace
       handle_view_on_rebase(diff_view_meta, /* indirect */ true);
-      TORCH_INTERNAL_ASSERT(diff_view_meta->output_nr_ == 0);
```
We don't want to do that I think. Many other things are going to break if you do that.
I think what you want to do here is:
- In this function, change the original condition to only do the rebase if it is a view and not MULTI_OUTPUT_SAFE.
- In the rebase_history function, similarly, only do the view handling if it is not MULTI_OUTPUT_SAFE.

This will ensure that for now, these function have the exact same behavior as before (as if they were not views + no version tracking) and we only add the warning.
Next version, we can remove this CreationMeta and add back the version tracking.

albanD reviewed (commented):
Hey,

Thanks a lot for sending this PR!
I would mostly have questions about the content of the lists of views and the behavior of the new functions. But it is mostly good I think.

Baranowski commented on tools/autograd/gen_autograd.py:
```diff
@@ -64,11 +64,23 @@
-RETURNS_VIEWS_OF_INPUT = set(VIEW_FUNCTIONS.keys()).union({'chunk', 'split'})
+RETURNS_VIEWS_OF_INPUT = set(VIEW_FUNCTIONS.keys()).union({
+    'chunk',
+    'split',
```
There is this comment above:
```
# note: some VIEW_FUNCTIONS are just compositions of the view functions above                            
# this list contains both the root view functions and any that are purely composed                       
# of viewing functions, and is used by the JIT to determine when an operator                             
# returns a view of its inputs
```

and in `aten/src/ATen/native/TensorShape.cpp` you can see that `split` calls `Tensor::narrow` so I thought that `self` belongs in `RETURNS_VIEWS_OF_INPUT` rather than `VIEW_FUNCTIONS`. But I must admit that I don't fully understand what's going on here.

albanD replied:
The actual implementation is "unrelated" here.
The short answer is that we have 2 ways to call a low level function:
- As a composite op where we have a single generatic implementation that calls into other ops
- As a non-composite op for which the dispatcher is responsible to call the right implementation.

These match the `elementary` vs `non-elementary` I mentioned above.
Finding which class each function belong to is a bit tricky though but in general:
- If there is an entry in derivatives.yaml, it is not composite
- If there are special type dispatch in native_functions.yaml, it is not composite

To be 100% sure, I usually check the `torch/csrc/autograd/generated/VariableTypeEverything.cpp` file and find my function in there. If the inner call is wrapped into `AutoNonVariableTypeMode` then it is not composite, otherwise it is.


For the case of `split`, you can see that it has an entry in `derivatives.yaml` so it should be in the first list.

Baranowski replied:
Thanks for the explanation. I will need to read it a few more times to understand, but your remark about  `torch/csrc/autograd/generated/VariableTypeEverything.cpp` provides easy enough rule for what to do here.

Baranowski commented:
@albanD I got stuck on something and I would appreciate your input:

I had to modify `nn.GRU` to use `unsafe_chunk` instead of `chunk`, as a result the onnx tests in CI fail because pytorch doesn't know how to export `unsafe_chunk` I wanted to simply add:

```
def unsafe_chunk(...):
  return chunk(...)
```

in `torch/onnx/symbolic_opset9.py` but there is no `chunk` function in any of the opset modules, I don't understand why. As a result, I also don't understand what's the right way to add support for `unsafe_chunk` in `torch.onnx`

neginraoof commented:
Hi @Baranowski 
ATen chunk op is actually handled within this jit pass: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/canonicalize_graph_fuser_ops.cpp

ONNX exporter uses this jit pass which replaces aten::chunk op with prim::ConstantChunk.
I believe we need a similar behavior for unsafe_chunk. You'd be able to find the symbolic function for prim_ConstantChunk in opset 9 symbolics.

Baranowski commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -482,6 +482,17 @@ def prim_ConstantChunk(g, self, chunks, dim):
     return prim_ConstantSplit(g, self, split_size, dim)
 
 
+@parse_args('v', 'i', 'i')
+def unsafe_chunk(g, self, chunks, dim):
+    split_size = (self.type().sizes()[dim] + chunks - 1) // chunks
+    size = self.type().sizes()[dim]
+    splits = [split_size] * (size // split_size)
+    leftover = size % split_size
+    if leftover:
+        splits.append(leftover)
+    return g.op("Split", self, split_i=splits, axis_i=dim, outputs=1)
```
I have instead added this `unsafe_chunk` which repeats the implementation of `prim_ConstantChunk` except that in the final statement it passes `outputs=1` instead of `outputs=len(splits)`. I think this is simpler than messing with jit pass and should still do the right thing. But it would be great if someone familiar with onnx could review my changes to this file.

albanD replied:
@neginraoof does that look good for you?

Baranowski committed:
import RETURNS_VIEWS_OF_INPUT

Baranowski committed:
Lint

albanD commented on tools/autograd/gen_autograd.py:
```diff
@@ -71,6 +72,14 @@
     VIEW_FUNCTIONS[key] = 'self'
 
+# Functions for which we use CreationMeta::MULTI_OUTPUT_SAFE. I.e., the ones for
+# which inplace modification of otputs is being gradually deprecated.
```
nit outputs

albanD commented on tools/autograd/gen_autograd.py:
```diff
@@ -45,10 +45,11 @@
     'as_strided': 'self',
     'diagonal': 'self',
     'expand': 'self',
-    'narrow': 'self',
```
You want to add it to `RETURNS_VIEWS_OF_INPUT` below

albanD commented on torch/csrc/autograd/variable.h:
```diff
@@ -377,9 +377,12 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
 /// - NO_GRAD_MODE should be set when a view in created when GradMode is disabled
 /// - MULTI_OUTPUT_NODE should be set when a Node created by codegen code returns
 ///   multiple differentiable views
+/// - MULTI_OUTPUT_SAFE should be set when a view was returned by a function
```
Could you add here: These are note considered as views for now to for the view+inplace logic! The graph won't be rewritten when an inplace is done, only a warning will be thrown.

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -79,22 +79,25 @@ namespace impl {
-      self.grad_fn(); // trigger an update to the view's grad_fn
-    } else {
-      set_gradient_edge(self, std::move(gradient_edge));
+      if (diff_view_meta->creation_meta != CreationMeta::MULTI_OUTPUT_SAFE) {
```
Can you add a comment pointing to the `[ View + Inplace detection ]` Note here and in the same if statement below?
To make it clear for future readers why these are not "real views" for now.

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -452,6 +457,12 @@ void handle_view_on_rebase(DifferentiableViewMeta* diff_view_meta, bool indirect
+      } else if (diff_view_meta->creation_meta == CreationMeta::MULTI_OUTPUT_SAFE) {
+        msg = c10::str(msg, " This view is the output of a function that "
+                       "returns multiple views. Inplace operators on such "
+                       "views are being deprecated and soon will not be "
```
I would be more aggressive here and say `v1.8` (given that this will go into v1.7 most likely.

albanD reviewed (commented):
For the backward compat, you just want to add `split_with_sizes` to the whitelist there: https://github.com/pytorch/pytorch/blob/a04af4dccb91eba013ca61da161ce857b5d6833b/test/backward_compatibility/check_backward_compatibility.py#L117
You are willingly changing the signature so it is fine.

The rest looks good. Just some comments in the code to prevent future readers to be super confused by this :)

Baranowski committed:
Alban's review

Baranowski commented:
I think I addressed all your remarks, Alban. Are we still waiting for a review/input from @neginraoof?

FYI I won't have the time to work on this after two weeks from now, so it would be best to merge this PR within that time.

albanD commented on tools/autograd/gen_autograd.py:
```diff
@@ -71,14 +72,22 @@
 
+# Functions for which we use CreationMeta::MULTI_OUTPUT_SAFE. I.e., the ones for
+# which inplace modification of outputs is being gradually deprecated.
+MULTI_OUTPUT_SAFE_FUNCTIONS = {
```
I don't think split and chunk need to be here right? They are just implemented with `split_with_sizes` and don't have any custom logic in them?

Baranowski replied:
I think that `chunk` indeed doesn't need to be here but `split` and `split_with_sizes` have separate implementations (that both call `narrow`), so I think those should remain.

albanD replied:
Ho very good point for `split`. But since we define a special derivative for it, we should move it from `RETURNS_VIEWS_OF_INPUT` to `VIEW_FUNCTIONS` in `tools/autograd/gen_autograd.py`.
This won't change the codegen now that you updated the generation to check both lists. But just to be consistent with the comments there.

albanD replied:
EDIT: You actually already added it in `VIEW_FUNCTIONS`, so it is in both right now haha

albanD commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -114,7 +115,7 @@
     ('aten::__or__', datetime.date(2020, 6, 30)),
     ('aten::__xor__', datetime.date(2020, 6, 30)),
-    ('aten::split', datetime.date(2020, 6, 30)),
+    ('aten::split', datetime.date(2020, 7, 20)),
```
This one you're not changing the signature, so you can actually just remove the line :D

albanD reviewed (commented):
Just one more nit I missed earlier and it should be good to go :D 

I think that if all the onnx test pass, it should be a good sign that it is the right fix!

Baranowski committed:
Alban's second review

albanD reviewed (approved):
LGTM !

albanD commented:
One interesting thing that internal testing brought up is that you can't do the following anymore:
```python
outs = inp.split_with_sizes(...)
res = outs[0].detach_()
```

This is definitely wrong but I think we want to make this a warning for now and forbid it later.
Can you update the implementation [here](https://github.com/pytorch/pytorch/blob/03eec07956a61cb824f4685ba3630f553b1a0a6d/torch/csrc/autograd/VariableTypeManual.cpp#L253) to only raise a warning for the safe views saying this is deprecated and will be an error soon?

Baranowski committed:
RETURNS_VIEWS_OF_INPUT -= 'split'

Baranowski committed:
Warning in detach_()

Baranowski committed:
Fix warning msgs

albanD commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -251,7 +251,18 @@ Tensor detach(const Tensor & self) {
+        TORCH_WARN("This view is an output of a function that "
+                   "returns multiple views. Detaching such views inplace "
+                   "is being deprecated and will be forbidden "
+                   "starting from version 1.8. Consider using `unsafe_` "
```
I think you want to recommend the out of place version here, there is no unsafe

Baranowski replied:
Done

Baranowski committed:
Recommend detach()

Baranowski committed:
Suggest not using inplace operators

albanD commented:
Unfortunately this got reverted as it lead to some internal issues (not sure yet if they are related or not). Looking into it and I will re-land it.
